pydantic/fields.py

```diff
@@ -705,6 +705,10 @@ def _apply_validators(
     ) -> 'ValidateReturn':
         for validator in validators:
             try:
+                # optional field set to None should be valid unless a custom validator handles this case
+                if validator.__name__ in DEFAULT_VALIDATORS_NAMES and v is None and not self.required:
```

Comment:
I'm not really satisfied with the check on name. But the validators are wrapped so no easy way to directly check that. I'm open to proposals

Reply:
Instead modify `ConstrainedList` to something like

```py
# This types superclass should be List[T], but cython chokes on that...
class ConstrainedList(list):  # type: ignore
    # Needed for pydantic to detect that this is a list
    __origin__ = list
    __args__: List[Type[T]]  # type: ignore

    min_items: Optional[int] = None
    max_items: Optional[int] = None
    item_type: Type[T]  # type: ignore

    @classmethod
    def __get_validators__(cls) -> 'CallableGenerator':
        yield cls.list_length_validator

    @classmethod
    def __modify_schema__(cls, field_schema: Dict[str, Any]) -> None:
        update_not_none(field_schema, minItems=cls.min_items, maxItems=cls.max_items)

    @classmethod
    def list_length_validator(cls, v: 'List[T]', field) -> Optional['List[T]']:
        if v is None and not field.required:
            return None

        v = list_validator(v)
        v_len = len(v)

        if cls.min_items is not None and v_len < cls.min_items:
            raise errors.ListMinLengthError(limit_value=cls.min_items)

        if cls.max_items is not None and v_len > cls.max_items:
            raise errors.ListMaxLengthError(limit_value=cls.max_items)

        return v
```

Reply:
> I'm afraid I'm not happy with this. This is the most performance critical part of the whole of pydantic.
> 
> I think we can fix this bug without modifying this funciton.

I definitely understand ! I was not satisfied either. But I wanted to tackle the whole problem instead of just the ConstrainList one. As a matter of fact it seems it's not needed! Sorry for the overkill